pkg/authz/check/service.go

```diff
@@ -405,92 +314,395 @@ func (svc CheckService) CheckMany(ctx context.Context, authInfo *service.AuthInf
 
 // Check returns true if the subject has a warrant (explicitly or implicitly) for given objectType:objectId#relation and context
 func (svc CheckService) Check(ctx context.Context, authInfo *service.AuthInfo, warrantCheck CheckSpec) (bool, []warrant.WarrantSpec, *wookie.Token, error) {
-	log.Ctx(ctx).Debug().Msgf("Checking for warrant %s", warrantCheck.String())
-
 	// Used to automatically append tenant context for session token w/ tenantId checks
 	if authInfo != nil && authInfo.TenantId != "" {
-		svc.appendTenantContext(&warrantCheck, authInfo.TenantId)
+		if warrantCheck.CheckWarrantSpec.Context == nil {
+			warrantCheck.CheckWarrantSpec.Context = warrant.PolicyContext{
+				"tenant": authInfo.TenantId,
+			}
+		} else {
+			warrantCheck.CheckWarrantSpec.Context["tenant"] = authInfo.TenantId
+		}
 	}
 
-	var match bool
-	decisionPath := make([]warrant.WarrantSpec, 0)
-	var newWookie *wookie.Token
-	newWookie, e := svc.WookieSvc.WookieSafeRead(ctx, func(wkCtx context.Context) error {
-		// Check for direct warrant match -> doc:readme#viewer@[10]
-		matchedWarrant, err := svc.getWithPolicyMatch(ctx, warrantCheck.CheckWarrantSpec)
+	// Fetch object types upfront
+	typesMap, _, err := svc.ObjectTypeSvc.GetTypeMap(ctx)
+	if err != nil {
+		return false, nil, nil, err
+	}
+
+	// TODO: Should do wookieSafeRead
+	// TODO: Set some 'dynamic' upper-bound for maxSubtaskConcurrency & cancelable-context
+	maxServiceConcurrency := 4
+	maxSubtaskConcurrency := 1000
```

Comment:
We should make these two values part of the overall config (i.e. values in`warrant.yaml`) and add some notes in the README about them